opentelemetry-dynatrace/src/transform/metrics.rs

```diff
@@ -0,0 +1,1165 @@
+                            Some(c)
+                        } else if !prev_char_underscore {
+                            prev_char_underscore = true;
+                            Some('_')
+                        } else {
+                            None
+                        }
+                    })
+                    .take(255)
+                    .collect::<String>();
+
+                if write!(acc, "{}", value).is_err() {
+                    acc.truncate(offset);
+                    return acc;
+                }
+
+                acc
+            });
+
+        fmt.write_str(&val)?;
+
+        Ok(())
+    }
+}
+
+impl PartialEq for DimensionSet {
+    fn eq(&self, other: &Self) -> bool {
+        self.dimensions.iter().eq(other.iter())
+    }
+}
+
+impl From<Vec<KeyValue>> for DimensionSet {
+    fn from(collection: Vec<KeyValue>) -> Self {
+        DimensionSet {
+            dimensions: collection
+                .into_iter()
+                .map(|kv| (kv.key, kv.value))
+                .collect(),
+        }
+    }
+}
+
+impl From<Vec<(Key, Value)>> for DimensionSet {
+    fn from(collection: Vec<(Key, Value)>) -> Self {
+        let mut dimensions = BTreeMap::new();
+        for (key, value) in collection.into_iter() {
+            dimensions.insert(key, value);
+        }
+        DimensionSet { dimensions }
+    }
+}
+
+impl FromIterator<KeyValue> for DimensionSet {
+    fn from_iter<I: IntoIterator<Item = KeyValue>>(iter: I) -> Self {
+        let mut dimensions = BTreeMap::new();
+        for kv in iter {
+            dimensions.insert(kv.key, kv.value);
+        }
+        DimensionSet { dimensions }
+    }
+}
+
+impl FromIterator<(Key, Value)> for DimensionSet {
+    fn from_iter<I: IntoIterator<Item = (Key, Value)>>(iter: I) -> Self {
+        let mut dimensions = BTreeMap::new();
+        for (key, value) in iter {
+            dimensions.insert(key, value);
+        }
+        DimensionSet { dimensions }
+    }
+}
+
+/// An iterator over the entries of a `DimensionSet`.
+#[derive(Debug)]
+pub struct Iter<'a>(btree_map::Iter<'a, Key, Value>);
+
+impl<'a> Iterator for Iter<'a> {
+    type Item = (&'a Key, &'a Value);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.0.next()
+    }
+}
+
+impl<'a> IntoIterator for &'a DimensionSet {
+    type Item = (&'a Key, &'a Value);
+    type IntoIter = Iter<'a>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        Iter(self.dimensions.iter())
+    }
+}
+
+/// [Dynatrace metrics ingestion protocol line].
+///
+/// [Dynatrace metrics ingestion protocol line]: https://www.dynatrace.com/support/help/how-to-use-dynatrace/metrics/metric-ingestion/metric-ingestion-protocol/
+#[derive(Clone, Debug)]
+pub struct MetricLine {
+    kind: NumberKind,
+    key: MetricKey,
+    dimensions: Option<DimensionSet>,
+    min: Option<Number>,
+    max: Option<Number>,
+    sum: Option<Number>,
+    count: Option<u64>,
+    delta: Option<Number>,
+    gauge: Option<Number>,
```

Comment:
For the payload part. Maybe using an enum here is more intuitive and will reduce the size by a little bit.

Reply:
Yes, I'm going to improve this in a subsequent PR. This thing doesn't look good at all and definitively needs some work.

Reply:
@diceride Can we a TODO item here? Thanks in advance